### PR TITLE
Access log interceptor

### PR DESCRIPTION
**Pull request by Tobi98**
Add access logging to log information about the request to standard output. The log contains information about the requester and the request itself. This can be useful for fulfilling legal requirements regarding data integrity or for forensics.

Example Log:
```sh
demo-api: [Nest] 30396  - 10/23/2023, 3:24:33 PM     LOG [AccessLogInterceptor] Request type: graphql | ip: ::1 | user: 1 (Test Admin) | operationType: Query | operationName: Test | resolver function: currentUser | args: {}
```

Closes COM-2

### CODE REVIEW

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+/** TODOs
+ * Docs are missing
+ * Api password check
+ */
+const DAM_IMAGE_PATH = "/dam/images/:hash/:fileId";
+const DAM_FILE_PATH = "/dam/files/:hash/:fileId";
+
+@Injectable()
+export class AccessLoggerInterceptor implements NestInterceptor {
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:12]: I'd call it AccessLogInterceptor

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import { CallHandler, ExecutionContext, Injectable, Logger, NestInterceptor } from "@nestjs/common";
+import { GqlExecutionContext } from "@nestjs/graphql";
+
+/** TODOs
+ * Docs are missing
+ * Api password check
+ */
+const DAM_IMAGE_PATH = "/dam/images/:hash/:fileId";
+const DAM_FILE_PATH = "/dam/files/:hash/:fileId";
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:9]: We also have  the preview URLs, which are used in the Admin previews. Does the logger ignore them too? Should it ignore them?

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        const additionalRequestData = [];
+        let ignored = false;
+
+        if (requestType === "graphql") {
+            const graphqlExecutionContext = GqlExecutionContext.create(context);
+            const graphqlContext = graphqlExecutionContext.getContext();
+
+            additionalRequestData.push(`ip: ${graphqlContext.req.ip}`);
+            const user = graphqlContext.req.user;
+            if (user) additionalRequestData.push(`user: ${user.id} (${user.name})`);
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:28]: This is the same for GraphQL and HTTP, could you extract it into a private function?

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            const gqlInfo = graphqlExecutionContext.getInfo();
+
+            if (gqlInfo.operation.operation === "mutation") {
+                delete gqlArgs["input"];
+                delete gqlArgs["data"];
+            }
+
+            message = `operationType: ${gqlInfo.parentType} - operationName: ${gqlInfo.operation.name.value} - resolver function: ${
+                gqlInfo.fieldName
+            } - args: ${JSON.stringify(gqlArgs)}`;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:40]: Do we need the resolver name? Shouldn't this be clear from the operation name?

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        } else {
+            const httpContext = context.switchToHttp();
+            const httpRequest = httpContext.getRequest();
+
+            if (httpRequest.route.path.includes(DAM_IMAGE_PATH) || httpRequest.route.path.includes(DAM_FILE_PATH)) ignored = true;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:45]: Code Style: I'd still prefer brackets for single line ifs

**Review by johnnyomair — COMMENTED** [review]
Nice first implementation!

Please also add a changeset 🙂

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts (hunk 2 above)
Comment (Tobi98) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:9]: thanks, i will check

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts (hunk 4 above)
Comment (Tobi98) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:40]: Yes it should, but then you always have to look explicitly into the admin code and find the resolver (if you can already see that the "error" happens in the api). if the resolver name is in the log you could jump directly into the api function.

But thats my opinion, I can also remove the resolver entry.

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+            additionalRequestData.push(`ip: ${httpRequest.ip}`);
+            const user = httpRequest.user;
+            if (user) additionalRequestData.push(`user: ${user.id} (${user.name})`);
+
+            message = `method: ${httpRequest.method} - route: ${httpRequest.route.path} - params: ${JSON.stringify(httpRequest.params)}`;
+        }
+        if (!ignored) {
+            this.logger.log(`Request type: ${requestType} {${additionalRequestData.join(" - ")}} - ${message}`);
```
Comment (fraxachun) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:54]: The formatting of the message seems a bit confusing to me, especially the usage of `{`...`}`. I'd prefer to make to either make it more readable (only use a single delimiter like `|`) or plain json for the whole line

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import { CallHandler, ExecutionContext, Injectable, Logger, NestInterceptor } from "@nestjs/common";
+import { GqlExecutionContext } from "@nestjs/graphql";
+
+/** TODOs
+ * Docs are missing
+ * Api password check
+ */
+const DAM_IMAGE_PATH = "/dam/images/:hash/:fileId";
```
Comment (fraxachun) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:8]: Better:
`const ignoredPaths = ["/dam/images/:hash/:fileId", "/dam/files/:hash/:fileId"];`

Comment (fraxachun) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:8]: You could also prepare something like `ignoredResolvers` to exclude common graphql-operations (like currentUser)

File: packages/api/cms-api/src/dam/files/entities/folder.entity.ts [hunk 8]
```diff
@@ -82,7 +82,7 @@ export function createFolderEntity({ Scope }: { Scope?: Type<DamScopeInterface>
         @Field()
         archived: boolean;
 
-        @Property({ columnType: "boolean" })
+        @Property({ columnType: "boolean", default: false })
```
Comment (fraxachun) [inline packages/api/cms-api/src/dam/files/entities/folder.entity.ts:85]: Shouldn't this be a separate PR?

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts (hunk 2 above)
Comment (Tobi98) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:9]: preview urls will not be ignored actually. Should we ignore them too?

File: packages/api/cms-api/src/dam/files/entities/folder.entity.ts (hunk 8 above)
Comment (Tobi98) [reply packages/api/cms-api/src/dam/files/entities/folder.entity.ts:85]: this change is not from me, maybe this chang is applied because @thomasdax98 changed the target branch to ubernext. I will update my PR to ubernext

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts (hunk 4 above)
Comment (nsams) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:40]: The operation name could be set so anything by an attacker, I would keep both

**Comment by fraxachun** [comment]
@Tobi98 Please mark resolved conversations as resolved.

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        } else {
+            const httpContext = context.switchToHttp();
+            const httpRequest = httpContext.getRequest();
+
+            if (
+                IGNORED_PATHS.some((ignoredPath) => httpRequest.route.path.includes(ignoredPath)) ||
+                (this.filterAuthenticatedRequest &&
+                    this.filterAuthenticatedRequest({
+                        user: httpRequest.user,
+                        authHeader: httpRequest.headers["authorization"] || httpRequest.headers["Authorization"],
```
Comment (fraxachun) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:64]: Is there a use case for this? The authorization header can contain anything, from jwt over opaque tokens to nothing depending on the current implementation. I think everything which could be necessary is part of the user-object above.

File: packages/api/cms-api/src/access-log/access-log.module.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import { APP_INTERCEPTOR } from "@nestjs/core";
+import { CurrentUser } from "src/user-permissions/dto/current-user";
+
+import { FILTER_AUTHENTICATED_REQUEST } from "./access-log.constants";
+import { AccessLogInterceptor } from "./access-log.interceptor";
+
+export type FilterAuthenticatedRequest = ({ user, authHeader }: { user?: CurrentUser; authHeader?: string }) => boolean;
+
+interface AccessLogModuleOptions {
+    filterAuthenticatedRequest?: FilterAuthenticatedRequest;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.module.ts:11]: I think we should consider adding an even more generic `filterRequests: (req: Req) => boolean` that allows us to filter any sort of request. Mostly this feature will be used to exclude request with the API_PASSWORD, but there may be even more specific requirements in some projects.

File: .changeset/olive-comics-knock.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+---
+"@comet/cms-api": minor
+"comet-docs": minor
+---
+
+Add AccessLog Interceptor
```
Comment (johnnyomair) [inline .changeset/olive-comics-knock.md:6]: Please be a little more specific: What is it, why do we add it, how can one integrate it into an application

File: packages/api/cms-api/src/access-log/access-log.module.ts (hunk 10 above)
Comment (fraxachun) [reply packages/api/cms-api/src/access-log/access-log.module.ts:11]: My suggestion: `filterRequests: (currentUser: CurrentUser, req: Req) => boolean`

**Comment by nsams** [comment]
Wouldn't `filterRquest` instead of `filterRequests` be more correct?

**Comment by nsams** [comment]
Index exports are missing - how did you test this at all? (I would add it to demo)

File: packages/api/cms-api/src/index.ts [hunk 12]
```diff
@@ -1,5 +1,6 @@
 import "reflect-metadata";
 
+export { AccessLogModule } from "./access-log/access-log.module";
```
Comment (Tobi98) [inline packages/api/cms-api/src/index.ts:3]: @nsams here is the index export

File: demo/api/src/app.module.ts [hunk 13]
```diff
@@ -154,6 +155,7 @@ export class AppModule {
                 PredefinedPageModule,
                 CronJobsModule,
                 ProductsModule,
+                AccessLogModule,
```
Comment (Tobi98) [inline demo/api/src/app.module.ts:158]: @nsams and here is the simplest implementation

**Review by nsams — DISMISSED** [review]

**Review by fraxachun — DISMISSED** [review]

**Comment by thomasdax98** [comment]
@Tobi98 I changed the branch because there is no breaking change, so we'll merge this to v5

File: .changeset/olive-comics-knock.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
+---
+"@comet/cms-api": minor
+"comet-docs": minor
+---
+
+Add AccessLog Interceptor where request data is logged in the standard console. The access log allows to see who accessed what data and when. This information for example can be used to identify security vulnerabilities.
```
Comment (dkarnutsch) [inline .changeset/olive-comics-knock.md:6]:
```suggestion
Add AccessLog Interceptor to log information about the request to standard output. The informations contains information about the requester and the request itself. This can be useful for fulfilling legal requirements regarding data integrity or for forensics.
```

File: docs/docs/auth/access-logging.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
+---
+title: Access Logging
+sidebar_position: 2
+---
+
+There is an option to enable global access logging, where request data is logged in the standard console. The following information is logged:
```
Comment (dkarnutsch) [inline docs/docs/auth/access-logging.md:6]:
```suggestion
There is an option to enable global access logging, where request data is logged in the standard output. The following information is logged:
```

File: docs/docs/auth/access-logging.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+-   For queries, the arguments (args)
+-   For mutations, the arguments excluding the input and data properties
+
+For HTTP requests, the following is logged:
+
+-   The HTTP method
+-   The route
+-   For GET requests, the parameters
+
+The access log allows to see who accessed what data and when. This information for example can be used to identify security vulnerabilities.
```
Comment (dkarnutsch) [inline docs/docs/auth/access-logging.md:27]:
```suggestion
The access log allows to see who accessed what data and when. This can be useful for fulfilling legal requirements regarding data integrity or for forensics.
```

File: .changeset/olive-comics-knock.md (hunk 14 above)
Comment (johnnyomair) [reply .changeset/olive-comics-knock.md:6]:
```suggestion
Add access logging to log information about the request to standard output. The log contains information about the requester and the request itself. This can be useful for fulfilling legal requirements regarding data integrity or for forensics.
```

The usage of an interceptor is an implementation detail IMO

File: .changeset/olive-comics-knock.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+There are two options for implementing the module in the app:
```
Comment (johnnyomair) [inline .changeset/olive-comics-knock.md:8]:
```suggestion
There are two ways to integrate logging into an application:
```

File: .changeset/olive-comics-knock.md [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+**1 option:**
```
Comment (johnnyomair) [inline .changeset/olive-comics-knock.md:10]:
```suggestion
**First option: Use the default implementation**
```

File: .changeset/olive-comics-knock.md [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+```ts
+imports: [
+    ...
+    AccessLogModule,
+    ...
+]
+```
+
+**2 option:**
```
Comment (johnnyomair) [inline .changeset/olive-comics-knock.md:20]:
```suggestion
**Second option: Configure logging**
```

File: .changeset/olive-comics-knock.md [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+With the second option, there is the opportunity to add a callback function called `filterRequest`. This function provides the ability to disable logging for various functionalities.
```
Comment (johnnyomair) [inline .changeset/olive-comics-knock.md:22]:
```suggestion
Use the `filterRequest` to prevent logging for specific requests. For instance, one may filter requests for system users.
```

File: docs/docs/auth/access-logging.md [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+## Configure AccessLog in app.module
```
Comment (johnnyomair) [inline docs/docs/auth/access-logging.md:29]:
```suggestion
## Configure access logging in the application
```

File: docs/docs/auth/access-logging.md [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+The optional function `filterRequest` provides the ability to disable logging for various functions. For example, requests executed during a build should not be logged. In this function, you have access to the user and the request object. If the function returns `true`, the log will not be executed for this condition.
```
Comment (johnnyomair) [inline docs/docs/auth/access-logging.md:31]:
```suggestion
The `filterRequest` callback can be used to prevent logging for specific requests. For example, requests executed during a build should not be logged. In this function, you have access to the user and the request object. If the function returns `true`, no log will be emitted for this request.
```

File: docs/docs/auth/access-logging.md [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+There are two options for implementing the module:
```
Comment (johnnyomair) [inline docs/docs/auth/access-logging.md:33]:
```suggestion
There are two ways to integrate logging into an application:
```

File: docs/docs/auth/access-logging.md [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+```ts
+imports: [
+    ...
+    AccessLogModule,
+    ...
+]
+```
+
+**2 option:**
```
Comment (johnnyomair) [inline docs/docs/auth/access-logging.md:45]:
```suggestion
**Second option: Configure logging**
```

File: docs/docs/auth/access-logging.md [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+The access log allows to see who accessed what data and when. This information for example can be used to identify security vulnerabilities.
+
+## Configure AccessLog in app.module
+
+The optional function `filterRequest` provides the ability to disable logging for various functions. For example, requests executed during a build should not be logged. In this function, you have access to the user and the request object. If the function returns `true`, the log will not be executed for this condition.
+
+There are two options for implementing the module:
+
+**1 option:**
```
Comment (johnnyomair) [inline docs/docs/auth/access-logging.md:35]:
```suggestion
**First option: Use the default implementation**
```

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+        if (requestType === "graphql") {
+            const graphqlExecutionContext = GqlExecutionContext.create(context);
+            const graphqlContext = graphqlExecutionContext.getContext();
+
+            if (
+                this.filterRequest &&
+                this.filterRequest({
+                    user: graphqlContext.req.user,
+                    req: graphqlContext.req,
+                })
```
Comment (johnnyomair) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:31]: Is it a good idea to disable logging when `filterRequests` returns `true`? This might be confusing, as `Array.filter` keeps the element in the array if the function returns `true`. Should we rename the function to be more clear? For instance `shouldLogRequest: () => boolean`?

Comment (Tobi98) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:31]: sounds like a good idea 👍🏼 

Another name could be `ignoreLogRequest: () => boolean`?
Consider the example of `shouldLogRequest`. If the function returns `true`, it might be misleading, as `true` suggests logging the request rather than ignoring it.

Comment (johnnyomair) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:31]: > If the function returns true, it might be misleading, as true suggests logging the request rather than ignoring it.

If the function returns true, it should log the request of course. The default implementation can be `() => true`

> Another name could be ignoreLogRequest: () => boolean?

IMO this is more confusing. To log a request, I'd need to return `false`. Seems like a double negative too me

Comment (nsams) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:31]: > it might be misleading, as true suggests logging the request rather than ignoring it.

I don't think this is misleading, it is very explicit what is happening.

Following all arguments, `shouldLogRequest: () => boolean` seems the best

Comment (dkarnutsch) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:31]: > shouldLogRequest: () => boolean seems the best

agree

Comment (fraxachun) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts:31]: In general I also like more the positive `shouldLogRequest` and even though I find the name not perfect I don't know a better one.

**Review by dkarnutsch — APPROVED** [review]

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts (hunk 26 above)
Comment (Tobi98) [reply packages/api/cms-api/src/access-log/access-log.interceptor.ts]: done in d0f7cdb9a4582de585dc80a9b167d9d7b3a27123

File: packages/api/cms-api/src/access-log/access-log.interceptor.ts [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+import { FILTER_AUTHENTICATED_REQUEST } from "./access-log.constants";
+import { FilterRequest } from "./access-log.module";
+
+const IGNORED_PATHS = ["/dam/images/:hash/:fileId", "/dam/files/:hash/:fileId", "/dam/images/preview/:fileId", "/dam/files/preview/:fileId"];
+
+@Injectable()
+export class AccessLogInterceptor implements NestInterceptor {
+    protected readonly logger = new Logger(AccessLogInterceptor.name);
+
+    constructor(@Optional() @Inject(FILTER_AUTHENTICATED_REQUEST) private readonly shouldLogRequest?: FilterRequest) {}
```
Comment (nsams) [inline packages/api/cms-api/src/access-log/access-log.interceptor.ts:14]: please name everything consistently (FILTER_AUTHENTICATED_REQUEST, FilterRequest)

**Review by johnnyomair — APPROVED** [review]

**Review by dkarnutsch — APPROVED** [review]

**Review by fraxachun — APPROVED** [review]
